cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -318,36 +337,58 @@ private void processMetrics() {
           topicMessageIds.put(topicIdMetaKey, currentMessageId);
         }
 
-        // Skip persisting if the records is empty
-        if (records.isEmpty()) {
-          return;
-        }
+        // Try to persist metrics and messageId's of the last metrics to be persisted if no other thread is persisting
+        tryPersist();
 
-        // Ensure there's only one thread can persist records and messageId's when the persistingFlag is false.
-        // The thread set persistingFlag to true when it starts to persist.
-        if (!persistingFlag.compareAndSet(false, true)) {
-          LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
-          return;
-        }
-        try {
-          Deque<MetricValues> recordsCopy = new LinkedList<>();
-          Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
-          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
-          Iterator<MetricValues> iterator = records.iterator();
-          while (iterator.hasNext()) {
-            recordsCopy.add(iterator.next());
-            iterator.remove();
-          }
-          persistRecordsMessageIds(recordsCopy, topicMessageIdsCopy);
-        } catch (Exception e) {
-          LOG.error("Failed to persist consumed messages.", e);
-        } finally {
-          // Set persistingFlag back to false after persisting completes.
-          persistingFlag.set(false);
+        long endTime = System.currentTimeMillis();
+        if (endTime - lastMetricTimeMillis > maxDelayMillis) {
```

Comment:
The `lastMetricTimeMillis` was not initialized. This mean if there is no messages being fetched, then `lastMetricTimeMillis` would be some old timestamp (potentially 0) and this is always true, resulting in polling TMS without sleep?